src/core/dense_set.cc

```diff
@@ -0,0 +1,558 @@
+  }
+
+  if (bid && entries_[bid - 1].IsEmpty()) {
+    return entries_.begin() + bid - 1;
+  }
+
+  return entries_.end();
+}
+
+bool DenseSet::Reserve(size_t sz) {
+  sz = std::min<size_t>(sz, kMinSize);
+
+  if (sz > entries_.max_size()) {
+    return false;
+  }
+
+  sz = absl::bit_ceil(sz);
+  capacity_log_ = absl::bit_width(sz);
+  entries_.reserve(sz);
+
+  if (entries_.capacity() != sz) {
+    return false;
+  }
+
+  return true;
+}
+
+void DenseSet::Grow() {
+  size_t prev_size = entries_.size();
+  entries_.resize(prev_size * 2);
+  ++capacity_log_;
+
+  for (long i = prev_size - 1; i >= 0; --i) {
+    ChainIterator prev = ChainEnd(i);
+    ChainIterator curr = ChainBegin(i);
+
+    while (curr != ChainEnd(i)) {
+      void* ptr = curr->GetObject();
+      uint32_t bid = BucketId(ptr);
+
+      if (bid == i) {
+        curr->ClearDisplaced();
+        prev = curr;
+        ++curr;
+      } else {
+        DensePtr node;
+        curr = Unlink(curr, prev, &node);
+        PushFront(entries_.begin() + bid, &node);
+        entries_[bid].ClearDisplaced();
+      }
+    }
+
+    if (prev != nullptr) {
+      DCHECK(!prev->IsLink());
+    }
+  }
+}
+
+bool DenseSet::AddInternal(void* ptr) {
+  uint64_t hc = Hash(ptr);
+
+  if (entries_.empty()) {
+    capacity_log_ = kMinSizeShift;
+    entries_.resize(kMinSize);
+    uint32_t bucket_id = BucketId(hc);
+    auto e = entries_.begin() + bucket_id;
+    obj_malloc_used_ += PushFront(e, ptr);
+    ++size_;
+    ++num_used_buckets_;
+
+    return true;
+  }
+
+  // if the value is already in the set exit early
+  uint32_t bucket_id = BucketId(hc);
+  if (FindAround(ptr, bucket_id).first != ChainEnd(bucket_id)) {
+    return false;
+  }
+
+  DCHECK_LT(bucket_id, entries_.size());
+
+  // Try insert into flat surface first. Also handle the grow case
+  // if utilization is too high.
+  for (unsigned j = 0; j < 2; ++j) {
+    ChainVectorIterator list = FindEmptyAround(bucket_id);
+    if (list != entries_.end()) {
+      obj_malloc_used_ += PushFront(list, ptr);
+      if (std::distance(entries_.begin(), list) != bucket_id) {
+        list->SetDisplaced(std::distance(entries_.begin() + bucket_id, list));
+      }
+      ++num_used_buckets_;
+      ++size_;
+      return true;
+    }
+
+    if (size_ < entries_.size()) {
+      break;
+    }
+
+    Grow();
+    bucket_id = BucketId(hc);
+  }
+
+  DCHECK(!entries_[bucket_id].IsEmpty());
+
+  /**
+   * Since the current entry is not empty, it is either a valid chain
+   * or there is a displaced node here. In the latter case it is best to
+   * move the displaced node to its correct bucket. However there could be
+   * a displaced node there and so forth. Keep a stack of nodes that are displaced
+   * until we find a displaced node which hashes to an empty bucket.
+   * Then unwind the stack of pending displacements until each node is in
+   * its correct bucket
+   */
+  std::stack<std::pair<DensePtr, ChainVectorIterator>> to_fix_displacements;
```

Comment:
are you sure that the stack is necessary?
can we keep just a single variable aside and do 
```
  while (!curr->IsEmpty() && curr->IsDisplaced()) {
    Unlink(curr, unlinked);
    pushfront(curr, ptr);
    ptr = unlinked.GetObject();   // get the next 
    curr -=  unlinked.GetDisplacedDirection();
}
```
i.e. you should always have a single element that is "out".
```

Reply:
yep this can be done with a single variable.